Step3 - 질문 삭제하기 리팩터링

lewisseo91 opened the pull request:
안녕하세요~!
리뷰 해 주셔서 정말 감사합니다!!
잘 부탁드립니다~!!

그리고 테스트 하면서 궁금했던 점이 있어 또 질문 드리게 되었습니다~

QuestionTest를 진행하면서 126line에
``` java
                () -> assertThat(actual.getCreatedAt()).isAfter(now),
                () -> assertThat(actual.getUpdatedAt()).isAfter(now)
```
위와 같이 assert 설정을 하고 만든 시간을 테스트 하고 있었습니다.
메소드 단일로 테스트할 때는 전혀 문제가 없었지만,
안에 있는 모든 메소드를 테스트 할 때는 계속 now 타임이 
createdAt과 updatedAt과 같은 타임으로 나왔습니다.
혹시 모든 메소드로 테스트할 때는 단일 메소드로 테스트할 때와 다르게 동작하게 되나요? 
관련 부분을 찾아보려고 했는데 이 부분과 관련된 찾을 키워드를 찾기가 어려워서 도움을 요청하게 되었습니다~!

lewisseo91 committed:
fix(Answer): answer 생성자 생성 시 question 연관 관계 매핑

answer 생성자 생성 시 question 연관 관계 매핑

lewisseo91 committed:
fix(docs): 순환 참조 관련 원인, 해결 방법 추가, 연관 관계 정리 추가

순환 참조 관련 원인, 해결 방법 추가, 연관 관계 정리 추가

lewisseo91 committed:
fix(docs): toString 제거, equals id만 남김

toString 제거, equals id만 남김
- view가 필요하면 dto 생성하는 식으로 변경할 것
- equals 는 id만 남겨도 된다고 생각해서 적용 (이 프로젝트에서는 엔티티만 보기 때문에 => 비영속 상태의 비교라고 한다면 좀 더 생각해 봐야 할 문제)

lewisseo91 committed:
fix(User): user <-> question 양방향 -> 단방향으로 수정

user <-> question 양방향 -> 단방향으로 수정
- 현재 요구사항에 필요한 기능도 아닌데 추가했다는 생각이 들어서 수정

lewisseo91 committed:
fix(Question): addAnswer 제거

answer 만들때 이미 question에서 추가하므로 제거

lewisseo91 committed:
fix(domain): hashcode 생성 시 순환 참조 에러

hashcode 생성 시 순환 참조 에러 (question, answer)
와 같은 에러를 생각해 id만으로 참조하도록 변경

lewisseo91 committed:
chore(User): 필요없는 import 정리

필요없는 import 정리

lewisseo91 committed:
refactor(QnaService): 서비스 로직 -> 도메인 로직 이동

Question.java
method 설명
- delete : question delete -> 내부 answer delete -> deleteHistory 목록 반환
- validateOwner : owner 확인
- deleteQuestion : question 삭제 후 deleteHistory 추가
- deleteAnswers : answers 삭제 후 deleteHistory 추가
Answer.java
- validateOwner : owner 확인
- deleteQuestion : answer 삭제 후 deleteHistory 반환
QnaService.java
- deleteQuestion 내부로 이동된 비즈니스 로직 삭제

lewisseo91 committed:
fix(QnaServiceTest): deleteQuestion 변경으로 인한 정리

addAnswer 필요없어서 제거 (생성자 생성 시에 들어감)
answerRepository findByQuestionIdAndDeletedFalse

lewisseo91 committed:
fix(domain): createdAt, updatedAt 테스트 생략

classTest 시에 time이 equal 하게 나오는 이슈로 인해 생략 시도
answerRepository.delete(answer) 후에 answer가 삭제되지 않는 이슈로 생략 시도

lewisseo91 committed:
refactor(Question): Question 문자열, 기본값 포장 추가

문자열, 기본값 포장 추가
- Answers
- QuestionTitle
- Contents
- Deleted
QuestionTest.java
- title 테스트 추가

lewisseo91 committed:
refactor(AnswerTest): repository.delete -> removeAnswer 로 변경

repository.delete -> removeAnswer 로 변경
isSameAs assert 테스트 추가

lewisseo91 committed:
refactor(Answer): Answer 원시값, 문자열 포장

Answer 원시값, 문자열 포장
- String contents -> Contents contents 변경
- boolean deleted -> Deleted deleted 변경

lewisseo91 committed:
refactor(DeleteHistory): DeleteHistory 원시값, 문자열 포장

DeleteHistory 원시값, 문자열 포장
- String contentId -> DeleteHistoryContentId contentId 변경

lewisseo91 committed:
refactor(User): User 원시값, 문자열 포장

User 원시값, 문자열 포장
- String userId -> UserId userId
- String password -> UserPassword password
- String name -> UserName name
- String email -> UserEmail email
로 변경
UserId, UserPassword, UserName, UserEmail 추가

lewisseo91 committed:
refactor(AnswerTest, QuestionTest): question delete, answer delete 테스트 추가

question delete with user, answer delete with user 테스트 추가
- question delete with user && with otherUser exception 추가
- answer delete with user && with otherUser exception 추가

lewisseo91 committed:
chore(QnaServiceTest): 필요없는 변수 제거

필요없는 변수 제거 (answerRepository)

lewisseo91 committed:
fix(UserEmail): embeddable 어노테이션 추가

embeddable 어노테이션 추가

lewisseo91 committed:
feat(QnaServiceTest): delete question 성공 시 answer 개수 테스트 추가

delete question 성공 시 answer 개수 테스트 추가

lewisseo91 committed:
fix(Answers): CascadeType ALL 로 변경

CascadeType ALL 로 변경
영속화 뿐만 아니라 제거나 refresh 할 때도 같이 관리하도록 시도

lewisseo91 committed:
chore(docs): CascadeType.ALL -> orphanRemoval true 했는데도 불구하고 repository 에서 delete 시도 시 쿼리 발생 안하는 이유

CascadeType.ALL -> orphanRemoval true 했는데도 불구하고 repository 에서 delete 시도 시 쿼리 발생 안하는 이유 추가

slamdunk7575 commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -35,24 +31,7 @@ public Question findQuestionById(Long id) {
-            answer.setDeleted(true);
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
+        List<DeleteHistory> deleteHistories = question.delete(loginUser);
         deleteHistoryService.saveAll(deleteHistories);
     }
```
깔끔하게 리팩토링 잘해주셨네요 👍

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -89,46 +90,47 @@ public Question getQuestion() {
     }
 
     public void setDeleted(boolean deleted) {
-        this.deleted = deleted;
+        this.deleted.setDeleted(deleted);
+    }
```
객체의 상태를 외부에서 변경하도록 드러내는 `set-` 이라는 네이밍 보다는
`예: delete()` 라는 이름을 주어 객체에게 메시지를 보내 처리되도록 나타내면 좋겠어요.

lewisseo91 replied:
상태 변경 메시지보다는 메시지를 보내야한다는 점을 신경써야겠네요~! 감사합니다~

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -89,46 +90,47 @@ public Question getQuestion() {
-                Objects.equals(writer, answer.writer) &&
-                Objects.equals(question, answer.question) &&
-                Objects.equals(contents, answer.contents);
+        return Objects.equals(id, answer.id);
     }
```
지난 내용도 잊지 않고 반영해보셨네요 💯

slamdunk7575 commented on src/main/java/qna/domain/Question.java:
```diff
@@ -92,51 +84,62 @@ public void setId(Long id) {
+
+    private void deleteQuestion(List<DeleteHistory> deleteHistories) {
+        setDeleted(true);
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, id, writer, LocalDateTime.now()));
```
`DeleteHistory` 내에 아래와 같은 `정적 팩토리 메소드`를 만들어보면
인자수도 줄면서 코드의 가독성을 높여줄 수 있겠네요 :)

```java
    public static DeleteHistory ofQuestion(Long contentId, User deletedBy) {
        return new DeleteHistory(ContentType.QUESTION, contentId, deletedBy);
    }

    public static DeleteHistory ofAnswer(Long contentId, User deletedBy) {
        return new DeleteHistory(ContentType.ANSWER, contentId, deletedBy);
    }
```

lewisseo91 replied:
팩토리메서드는 정말 활용도가 높은 패턴인 거 같습니다 ^^!!
감사합니다~

slamdunk7575 commented on src/main/java/qna/domain/QuestionTitle.java:
```diff
@@ -0,0 +1,35 @@
+import javax.persistence.Embeddable;
+import java.util.Objects;
+
+@Embeddable
+public class QuestionTitle {
+    @Column(nullable = false, length = 100)
+    private String title;
```
원시값 포장 객체 좋습니다 👍 
`title`에 대한 검증을 해볼 수도 있을것 같아요.

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -89,46 +90,47 @@ public Question getQuestion() {
+        this.deleted.setDeleted(deleted);
+    }
+
+    public DeleteHistory delete(User loginUser) throws CannotDeleteException {
+        validateOwner(loginUser);
+        return deleteAnswer();
+    }
```
리팩토링을 진행하면서 Service 의 비지니스 로직이 Domain 으로 옮겨졌는데요. 
필요하다고 생각되는 `도메인에(예: Answers) 단위 테스트`를 해보면 좋을 것 같아요.
기존 테스트를 `Repository`와 `Domain` 테스트로 나눠보는 것도 좋겠네요.

slamdunk7575 commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -121,12 +122,29 @@ void removeQuestionWithAnswerTest() {
-                () -> assertThat(actual.getCreatedAt()).isAfter(now),
-                () -> assertThat(actual.getUpdatedAt()).isAfter(now)
+                () -> assertThat(actual.getAnswers().size()).isZero()
+//                () -> assertThat(actual.getCreatedAt()).isAfter(now), // 질문하기!
+//                () -> assertThat(actual.getUpdatedAt()).isAfter(now)
```
> 위와 같이 assert 설정을 하고 만든 시간을 테스트 하고 있었습니다.
메소드 단일로 테스트할 때는 전혀 문제가 없었지만,
안에 있는 모든 메소드를 테스트 할 때는 계속 now 타임이
createdAt과 updatedAt과 같은 타임으로 나왔습니다.
혹시 모든 메소드로 테스트할 때는 단일 메소드로 테스트할 때와 다르게 동작하게 되나요?

전체 테스트를 돌려봤을 때 이상없이 통과를 하는데요 🤔
`fail` 나는 상황을 조금 자세히 설명해주시면 좋을 것 같아요. 캡쳐를 해주셔도 좋습니다.

lewisseo91 replied:
넵~! 이 쪽 부분은 제가 준비해서 올린 후에 코드 수정 후에 재요청 시에 스크린샷과 함께 올려보겠습니다 ㅎㅎ

slamdunk7575 reviewed (changes requested):
QnAService 리팩토링 깔끔하게 잘해주셨네요 👍
간단한 코멘트 몇가지 남겨드렸는데 확인해보시면 좋을 것 같아요.
끝까지 화이팅 입니다~!

slamdunk7575 commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,36 @@
+@Embeddable
+public class Answers {
+    @OneToMany(mappedBy = "question", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
+    private final List<Answer> answers = new ArrayList<>();
```
일급 컬렉션으로 분리를 잘해주셨네요 👍 

`@Where`를 활용하면 Question을 통해 삭제되지 않은 Answer를 가지고 올 수도 있습니다. 참고만 해주세요~!
`@Where(clause = "deleted = false")`

lewisseo91 committed:
refactor(domain): set~ 네이밍 -> 메시지 네이밍으로 변경

Answer.java
setContents -> changeContents
setDeleted -> 삭제
this.deleted.setDeleted -> this.deleted.delete
Question.java
setContents -> changeContents
this.contents.setContents -> this.contents.changeContents
setId -> 삭제
User.java
setEmail -> changeEmail
this.email.setEmail -> this.email.changeEmail

lewisseo91 committed:
refactor(DeleteHistory): 정적 팩토리 메서드 추가

정적 팩토리 메서드 추가, 적용

lewisseo91 committed:
refactor(Test): domainTest, RepositoryTest 로 나눔

domainTest, RepositoryTest 로 나눔

lewisseo91 committed:
feat(BaseEntity): createdAt, updatedAt getter 추가

createdAt, updatedAt getter 추가

lewisseo91 committed:
feat(domain, test): domain 검증 추가

Answers, QuestionTitle, UserEmail, UserId, UserName, UserPassword
- domain validate, save, validate test 추가

lewisseo91 commented:
안녕하세요~! 
정적 팩토리 메서드를 쓰면서 제가 ContentType을 잘못 쓴 것도 알고 하면서 얼마나 powerful 한 지 새삼 깨달았습니다.
좋은 피드백 정말 감사합니다!!

개인적 기준으로 RepositoryTest와 DomainTest를 나누어 봤습니다. 
then의 값이 Repository를 검증하면 RepositoryTest, Domain을 검증한다면 DomainTest를 하도록 고려해서 넣어보았습니다.

그리고 저번에 말씀드렸던 createdAt, updatedAt은 다시 하니 제대로 잘 나옵니다~!
실패를 어떻게 한 거지 생각하면서 이것저것 다 해봤는데도 제대로 잘 나옵니다.
혹시나 다음 미션에서 같은 문제가 생기면 물어보는 수밖에 없는 것 같습니다~!

@Where는 참고용으로 미션 종료 후에 테스트 해 보겠습니다 ㅎㅎㅎ

좋은 하루 되시길 바라요~!!

slamdunk7575 commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -21,26 +22,36 @@
     @Enumerated(EnumType.STRING)
-    private ContentType contentType;
+    private final ContentType contentType;
 
-    private Long contentId;
+    @Embedded
+    private final DeleteHistoryContentId contentId;
 
     @ManyToOne
     @JoinColumn(name = "deleted_by_id", foreignKey = @ForeignKey(name = "fk_delete_history_to_user"))
-    private User deletedBy;
+    private final User deletedBy;
```
`기본 생성자`를 놓치신 것 같네요.
생성자를 추가하면 final 키워드에 빨간불이 들어올텐데요.

JPA는 엔티티를 생성할 때 `지연로딩` 방식을 사용합니다.
이때 `프록시 객체`(PersistentBag)를 생성하여 사용하기 때문에 필드 또는 클래스에 `final`을 사용할 수 없습니다.

lewisseo91 replied:
앗 감사합니다!!

slamdunk7575 commented on src/main/java/qna/domain/Question.java:
```diff
@@ -71,11 +63,11 @@ public boolean isOwner(User writer) {
     public void addAnswer(Answer answer) {
-        answers.add(answer);
+        answers.addAnswer(answer);
         answer.setQuestion(this);
```
피드백 반영 잘해주셨네요 👍
여기도 `set-` 보다는 `예: toQuestion()`은 어떨까요?

lewisseo91 replied:
좋은 네이밍 감사합니다~!

slamdunk7575 commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -55,88 +54,49 @@ void saveAnswerTest () {
     void removeQuestionWithAnswerTest() {
         answer.removeQuestion();
-        Question questionFromRepo = questionRepository.findById(question.getId())
+        Question actual = questionRepository.findById(question.getId())
                 .orElseThrow(NoSuchElementException::new);
         assertAll(
-                () -> assertThat(questionFromRepo.getContents()).isEqualTo("contents1"),
-                () -> assertThat(questionFromRepo.getAnswers().size()).isZero(),
-                () -> assertThat(questionFromRepo.getCreatedAt()).isAfter(now),
-                () -> assertThat(questionFromRepo.getUpdatedAt()).isAfter(now)
+                () -> assertThat(actual.getContents()).isEqualTo("contents1"),
+                () -> assertThat(actual.getAnswers().size()).isZero()
         );
     }
```
테스트 작성시 `given - when - then` 에 맞춰 줄바꿈(+주석)을 통해 가독성을 높여주면 
나중에 동현님 또는 동료 개발자가 유지보수할 때 더 좋을 것 같네요. 
남은 미션에서 연습해보시면 좋겠어요 😄

```java
    @DisplayName("question remove with answer 삭제")
    @Test
    void removeQuestionWithAnswerTest() {
        // given
        final Answer answer = answerRepository.save(new Answer(user, question, "Answers Contents1"));

        // when
        question.removeAnswer(answer);
        Question actual = questionRepository.findById(question.getId())
                .orElseThrow(NoSuchElementException::new);

        // then
        assertAll(
                () -> assertThat(actual.getWriter()).isEqualTo(user),
                () -> assertThat(actual.getContents()).isEqualTo("contents1"),
                () -> assertThat(actual.getAnswers().size()).isZero(),
                () -> assertThat(actual.getCreatedAt()).isAfter(now),
                () -> assertThat(actual.getUpdatedAt()).isAfter(now)
        );
    }
```

lewisseo91 replied:
꼭 필요한 주석이네요! 감사합니다~!!

slamdunk7575 reviewed (changes requested):
동현님~ 피드백 반영 잘해주셨네요 👍
한가지 코멘트 남겨드렸는데 확인 부탁드립니다.

lewisseo91 committed:
fix(deleteHistory): DeleteHistory 기본 생성자 추가

DeleteHistory 기본 생성자 추가
- 엔티티 final은 지연로딩 방식이라 프록시 객체를 생성하여 사용하기 때문에 사용 불가

lewisseo91 committed:
fix(Question): setQuestion -> toQuestion

toQuestion으로 네이밍 변경

lewisseo91 committed:
fix(test): given, when, then 줄 바꿈 넣기

given, when, then 줄 바꿈 넣기

lewisseo91 commented:
안녕하세요~! 
리뷰어님 말씀 듣고 줄 바꿈 넣으면서
리뷰어님이 처음에 코드 봤을 때 읽기 정말 힘들었겠다고 반성했습니다. 죄송합니다!
세심한 리뷰 정말 감사합니다~!!

아래는 embedded 선언에도 final 키워드가 있었어서 같이 제거 하였습니다.
제거 고민을 하다가 [https://stackoverflow.com/questions/39415773/embedded-beans-as-final-attribute](https://stackoverflow.com/questions/39415773/embedded-beans-as-final-attribute)
여기를 보고 jpa spec 을 따르기로 생각하였습니다.

lewisseo91 committed:
fix(domain): final 제거

final 제거
https://stackoverflow.com/questions/39415773/embedded-beans-as-final-attribute

slamdunk7575 commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -22,17 +22,20 @@
     private LocalDateTime createDate = LocalDateTime.now();
 
+    protected DeleteHistory() {
+    }
```
피드백 반영 잘해주셨네요 💯

slamdunk7575 commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -127,8 +106,22 @@ void removeQuestionWithAnswerTest() {
+    @DisplayName("validate owner exception")
+    @Test
+    void removeQuestionWithUserExceptionTest() {
+        assertThatThrownBy(() -> {
+            // when
+            final User otherUser = userRepository.save(UserTest.LEWISSEO);
+            question.delete(otherUser);
+
+            // then
+        }).isInstanceOf(CannotDeleteException.class);
```
동현님이 필요하다고 생각되는 부분에 `줄바꿈 or 주석`을 주시면 됩니다 👍

```java
assertThatThrownBy(() -> {
            // when & then
            final User otherUser = userRepository.save(UserTest.LEWISSEO);
            question.delete(otherUser);
}).isInstanceOf(CannotDeleteException.class);
```

slamdunk7575 reviewed (approved):
마지막까지 열심히 해주셔서 감사합니다. 그동안 고생 많으셨어요. 🙇
미션이 종료 되더라도 다른 사람 코드와 비교해보면서 장단점을 비교해본다면
또 새로운 인사이트를 얻으실 수 있을꺼에요!
남은 미션도 꾸준히 진행하시고 끝까지 완주하시길 응원하겠습니다! 🎉

lewisseo91 commented:
리뷰어님 덕분에 많은 인사이트 얻었습니다!!
좋은 리뷰 해주셔서 정말 정말 감사합니다!!
